ibis/bigquery/tests/test_compiler.py

```diff
@@ -76,3 +76,100 @@ def test_to_timestamp(alltypes, timezone):
+FROM `ibis-gbq.testing.functional_alltypes`"""  # noqa: E501
+    assert result == expected
+
+    w3 = ibis.window(preceding=(4, 2),
+                     group_by='year', order_by='timestamp_col')
+    expr = t.mutate(win_avg=t.float_col.mean().over(w3))
+    result = expr.compile()
+    expected = """\
+SELECT *,
+       avg(`float_col`) OVER (PARTITION BY `year` ORDER BY `timestamp_col` ROWS BETWEEN 4 PRECEDING AND 2 PRECEDING) AS `win_avg`
+FROM `ibis-gbq.testing.functional_alltypes`"""  # noqa: E501
+    assert result == expected
+
+
+def test_range_window_function(alltypes):
+    t = alltypes
+    w = ibis.range_window(preceding=1, following=0,
+                          group_by='year', order_by='month')
+    expr = t.mutate(two_month_avg=t.float_col.mean().over(w))
+    result = expr.compile()
+    expected = """\
+SELECT *,
+       avg(`float_col`) OVER (PARTITION BY `year` ORDER BY `month` RANGE BETWEEN 1 PRECEDING AND CURRENT ROW) AS `two_month_avg`
+FROM `ibis-gbq.testing.functional_alltypes`"""  # noqa: E501
+    assert result == expected
+
+    w3 = ibis.range_window(preceding=(4, 2),
+                           group_by='year', order_by='timestamp_col')
+    expr = t.mutate(win_avg=t.float_col.mean().over(w3))
+    result = expr.compile()
+    expected = """\
+SELECT *,
+       avg(`float_col`) OVER (PARTITION BY `year` ORDER BY `timestamp_col` RANGE BETWEEN 4 PRECEDING AND 2 PRECEDING) AS `win_avg`
+FROM `ibis-gbq.testing.functional_alltypes`"""  # noqa: E501
+    assert result == expected
+
+
+@pytest.mark.parametrize(
+    ('preceding', 'value'),
+    [
+        (5, 5),
+        (ibis.microsecond(), 1),
+        (ibis.second(), 1000000.0),
+        (ibis.minute(), 1000000.0 * 60),
+        (ibis.hour(), 1000000.0 * 60 * 60),
+        (ibis.day(), 1000000.0 * 60 * 60 * 24),
+        (2 * ibis.day(), 1000000.0 * 60 * 60 * 24 * 2),
+        (ibis.week(), 1000000.0 * 60 * 60 * 24 * 7),
+    ]
+)
+def test_trailing_time_window(alltypes, preceding, value):
+    t = alltypes
+    w = ibis.trailing_time_window(preceding=preceding,
+                                  order_by=t.timestamp_col)
+    expr = t.mutate(win_avg=t.float_col.mean().over(w))
+    result = expr.compile()
+    expected = """\
+SELECT *,
+       avg(`float_col`) OVER (ORDER BY UNIX_MICROS(`timestamp_col`) RANGE BETWEEN {} PRECEDING AND CURRENT ROW) AS `win_avg`
+FROM `ibis-gbq.testing.functional_alltypes`""".format(value)  # noqa: E501
+    assert result == expected
+
+
+@pytest.mark.parametrize(
+    ('preceding', 'value'),
+    [
+        (ibis.year(), None),
+    ]
+)
+def test_trailing_time_window_unsupported(alltypes, preceding, value):
+    t = alltypes
+    w = ibis.trailing_time_window(preceding=preceding,
+                                  order_by=t.timestamp_col)
+    expr = t.mutate(win_avg=t.float_col.mean().over(w))
+    with pytest.raises(KeyError):
```

Comment:
Can we present a better error message here? It's not clear why this would generate a `KeyError` versus a `ValueError` that says "unsupported interval type". The fact that you're using a dictionary to look something up somewhere is unrelated to the error here.